### ISSUE TITLE
Recreating IC50 prediction

### ISSUE THREAD

**Issue by barabadwan** [PaccMann/paccmann_predictor#17]
Hi thanks for the prompt reply on the previous issue: I checked out the old version from git but now I run into an "index out of reach" error when I pass a SMILESlanguage into test_dataset creator, If I leave the Smiles language blank the code runs but the predictions seem off. See both attached images for code and results
![image](https://user-images.githubusercontent.com/16927097/152229016-65f017e0-cba3-42c1-a7b6-02ec61094832.png)
![image](https://user-images.githubusercontent.com/16927097/152230108-5411b68d-8f85-4853-8b1c-97eb9642bb68.png)

**Comment by jannisborn** [comment]
Hi @barabadwan,
Indeed, if you recreate the language the vocab will differ and the predictions would be more or less garbage. 
The script you are using is for model training, not for out-of-the-box testing the model on your own data. 
Anyways, a possible reason might be an incorrect pytoda version. Did you set up the env using the `conda.yml` in the `IC50` examples folder? The correct version should be in there. Also, the smiles language filepath should point to the smiles language object that is saved _within the model folder_

If you want to have out-of-the-box predictions of PaccMann, we recommend using the webservice: https://ibm.biz/paccmann-aas
For usage instructions, you can find info in the paper in NAR: https://academic.oup.com/nar/article/48/W1/W502/5836770?login=true

**Comment by barabadwan** [comment]
Yes installed pydota from the config file. Also, I am not using it to predict my own data, rather using the data you provided as you can see. Could it be a a mismatch of the padding var in the associated param files? 
![image](https://user-images.githubusercontent.com/16927097/152259012-b165c107-72cf-4a9c-b31e-b6c48c71184c.png)
![image](https://user-images.githubusercontent.com/16927097/152259470-043f150f-05b0-46b5-b6e2-44f142d6c150.png)

**Comment by jannisborn** [comment]
Indeed, the problem could be that the padding_size in your config file does not match the one that is expected by the model.  Or maybe the padding size is ok, but at inference time you encounter a SMILES that is larger than the given padding size. In this initial version of the model (which is the one that we released as pretrained), we did not really handle all those things perfectly.
How about retraining a PaccMann model using the latest version of the codebase + the data that we share? Especially the SMILES language was refactored a lot

**Comment by barabadwan** [comment]
When I try doing that the first epoch takes a long time to run and then pearsonr throws and error saying "expected data to be in tensor form". Is there is another issue that discusses how to resolve that? To be honest though it is a bit difficult to retrain the whole model. Would it be possible to upload a version of the pretrained and the model params that would work with the data provided?

**Comment by jannisborn** [comment]
Thanks @barabadwan I will look into this in the next days

**Comment by barabadwan** [comment]
Thank you, I was actually debugging some more and I realized the root of the error. It seems that defining the test_dataset using DrugSensitivityDataset actually changes the smiles_language that is passed into it. As you can see from the first screenshot the initial vocabulary size was 87 but after running DrugSensitivityDataset it changes to 93. I tried moving the params update to after the test_dataset is defined and now we manage to avoid the first error. However, the saved model weights are expecting there to be only 87 tokens not 93. Is it possible to stop the smiles language from being changed by DrugSensitivityDataset? Alternatively are there pretrained model weights that expect 93 tokens? 
Thanks
![image](https://user-images.githubusercontent.com/16927097/152415918-e1214420-913a-49bb-a09b-205f96d6e814.png)
![image](https://user-images.githubusercontent.com/16927097/152415972-bd35a4c0-39db-4213-ac5e-83b17a7e26ec.png)

**Comment by jannisborn** [comment]
In general this behavior occurs because the intended code logic was that **either** no smiles language is provided (in which case it is created on the fly. This should be done in training only) **or** the smiles language is provided **and** it resonates with the dataset (so the dataset does not introduce new tokens).
Your usecase is a bit in between since you want to use an existing language but still use it on data that introduces new tokens. 

For development, the two cases mentioned above were fine, but I completely agree that our code should handle this better and, indeed, in the current version of pytoda this is done seamlessly. However, because you are using the old pretrained model, you are bound to this old version. 

In this old version there is no native way to prevent the smiles language from being updated. But I recommend as follows:
- check which are the 6 new tokens (you can do `smiles_language.index_to_token`) and the last 6 entries will show the newly added tokens. 
- Do they occur frequently in your dataset and are they critical? 
- If not, you can uncomment lines 156-159 in `pytoda.datasets._smiles_dataset`: https://github.com/PaccMann/paccmann_datasets/blob/d25855b93e7e9111441fe4a2558e855edc9db6bc/pytoda/datasets/_smiles_dataset.py#L156 It's a hack but will stop adding new tokens to the language. At runtime, these tokens will be _ignored_ in the creation of the embedding matrices.
-  Note that uncommenting this lines will also stop adjusting the max_token_sequence_length. This is important for the padding, but it only matters for you if your dataset has molecules with more than 1261 tokens (I'd be surprised)

Lmk whether this suffices for you or you want me to still look into having an updated model. I might find one somewhere but not 100% sure

**Comment by barabadwan** [comment]
So commenting out those lines have allowed the model to run but the results aren't reflective of what you report in the paper, see attached. 

You keep referring to my use case as being somewhat different or that I am using my own data but I am simply using the same data files that were provided.  I am trying to predict the GDSC test set that was processed and submitted by your group and described in the README.md 

> test_sensitivity_filepath=input_dir+'data/splitted_data/gdsc_cell_line_ic50_test_fraction_0.1_id_997_seed_42.csv'
> gep_filepath=input_dir+'data/gene_expression/gdsc-rnaseq_gene-expression.csv'
> gene_filepath=input_dir+'data/2128_genes.pkl'
> smi_filepath=input_dir+'data/smiles/gdsc.smi'
> smiles_language_filepath=input_dir+'data/smiles_language_chembl_gdsc_ccle.pkl'
> model_weights = input_dir+"data/trained_models/paccmann/weights/best_rmse_mca.pt"
> params_filepath = input_dir+"data/trained_models/paccmann/model_params.json"
> model_dir = input_dir+"data/trained_models/paccmann/"

So I don't understand why the SMILESlanguage (smiles_language_chembl_gdsc_ccle) doesn't resonate with the dataset (gdsc_cell_line_ic50_test_fraction_0.1_id_997_seed_42) as it is the same one that was used to conduct the training. I am not introducing my own in data in any way, I am simply recreating the prediction that was already done with the code provided.


![image](https://user-images.githubusercontent.com/16927097/152439301-e57610d2-6fc0-4a6e-a3b7-7f78fb4b9942.png)

**Comment by jannisborn** [comment]
Hi @barabadwan,
sorry for the confusion but indeed, the original implementation is in tensorflow, see here: https://github.com/drugilsberg/paccmann
I recommend to check here:  https://ibm.biz/paccmann-data, there you will find the data split exactly like for the results reported in the paper (25-fold CV for strict split, 5-fold CV for lenient split). There's also a results folder which lists the predictions for all samples of the folds and allows you to recreate Figure 4 of the original paper.

**Comment by barabadwan** [comment]
How about this, are you able to run the train paccmann script that is listed in the examples folder and upload the model weights you get? I think that should probably resolve the issue

**Comment by jannisborn** [comment]
Hi @barabadwan!
I just trained a new PaccMann model and uploaded it here: https://ibm.biz/paccmann-data (see the `models` subfolder). During training, I got a Pearson correlation of 0.895 on the test data. There's also a README that explains how to use PaccMann on your own data or how to reproduce the results.

The code for the evaluation is on this repo, but you have to `git checkout dev` because it's not yet merged into master.
Note that this is a single model, so the performance will be slightly below what we found in the paper with the ensemble. 
For details, see the README in the model folder.
Hope this helps, let me know!

**Comment by jannisborn** [comment]
Hi @barabadwan,
did this solution work for you?

**Comment by barabadwan** [comment]
Unfortunately not, seems to have problem loading the model 
![image](https://user-images.githubusercontent.com/16927097/155231279-89d04288-bf0c-4951-a03a-ee93af0e18d2.png)

**Comment by jannisborn** [comment]
Unfortunately, the screenshot is not very informative on what is the problem here.
Did you setup the environment correctly, see details in the README:

```sh
git clone https://github.com/PaccMann/paccmann_predictor
cd paccmann_predictor
git checkout dev
conda env create -f examples/IC50/conda.yml
conda activate paccmann_predictor
pip install -e .
```

EDIT: Afterwards do:
```sh
pip install torch==1.7.0
```

**Comment by barabadwan** [comment]
Yes, but to be clear, the venv specified in the conda.yml is paccmann_predictor_dev. I also downloaded the specific versions of pytoda and torch

**Comment by jannisborn** [comment]
No, the environment name specified in the file is `paccmann_predictor` if you follow the instructions.

I think your error is related to the pytorch version. If you upgrade it to 1.7.0 (or higher), it should work (see my edit above)

**Comment by barabadwan** [comment]
Yes updating pytorch did the trick, thank you! 
I thought I had followed the installation instructions and the conda.yml file had specified pytorch 1.0.1 but glad things worked

**Comment by barabadwan** [comment]
Hi Jannis, thanks for all your help, I have a quick question about the results from this model. In the paper I see that sometimes the output prediction correlates directly with the IC50 so a lower number means its a more effective drug, and at some other points the prediction seems to be directly proportional to efficacy. As shown in the two graphs from one of your papers. I am wondering if the model output is the IC50 or is it somehow inverted and normalized.
<img width="913" alt="image" src="https://user-images.githubusercontent.com/16927097/164453300-e62dcfd4-28d2-4a02-846e-cebc89820a37.png">
<img width="837" alt="image" src="https://user-images.githubusercontent.com/16927097/164454045-f06a4ca2-2e86-43be-b4ab-d65f1988f206.png">

**Comment by jannisborn** [comment]
Hi @barabadwan,

thanks for the question. The general answer to this is that it hinges upon (1) the data that was provided and (2) the training process of the model. 

1. First thing should be to check the training data and the exact metric/scale that was used in there. These days, we're usually working with pIC50 in the data files but many of the older models used data in micromolar or other scales.
2. Here I would check the config files of the trained model. There is a parameter called `drug_sensitivity_min_max`. If set to True, the IC50 values are normalized to `[0,1]`, so you would need to de-normalize the model predictions first.

**Comment by jannisborn** [comment]
The specific answer is that the pretrained model I released for you above was trained on the GDSC data that we also released on Box (`gdsc_cell_line_ic50_train_fraction_0.9_id_997_seed_42.csv`).

In this file, the IC50 data is in `log10(micromolar IC50)`. So to go to pIC50 you need to do:
```py
to_pic50 = lambda x: np.log10(np.exp(x)*10**6)
```
imho this is the scale that is the easiest to interpret, but it's up to you how to transform

**Comment by barabadwan** [comment]
Great thank you, and I am looking at the param file you sent the drug_sensitivity_min_max is set to false so shouldn't be a problem there either

**Comment by barabadwan** [comment]
Sorry a little confused about this, wouldn't the transformation you suggest keep the ordering the same? So something with a low  log10(micromolar IC50) ie effective drug, would end up with a low pIC if we simply take  to_pic50 = lambda x: np.log10(np.exp(x)*10**6)? Seems like the original data might be -log10(micromolar IC50)

**Comment by jannisborn** [comment]
yes you are right there was a minus in the original data
